### ISSUE TITLE
adding per-feature event listeners

### ISSUE THREAD

**Issue by ansis** [mapbox/mapbox-gl-js#6215]
A list of questions about events and what I think might be the answers. What do you think?

## Per-feature event listeners?

Currently you can only listen to events on layers. I think we agree that we should expand this to individual features. The clearest advantage of this is with `mouseenter` and `mouseleave` events. Without per-feature event listeners you would need to manually track the features under the pointer and diff them.

## Dispatched once as a batch or once per feature?
We could dispatch a single `click` event with all the features under the pointer. We could dispatch a `mouseenter` event with all the features that were entered. But this seems weird. 

I think we should dispatch a callback once per feature. I'm having trouble figuring out *why*, but maybe because it's closer to how platforms handle things.

## What order should events be dispatched?

Top-down seems like the logical order. I think events on layers would be dispatched just below features in that layer.

```
register listener on all features in layerA
register listener on layerA
register listener on all features in layerB
click
dispatch click on feature1 in layerA
dispatch click on feature2 in layerA
dispatch click on layerA
dispatch click on feature5 in layerB
```

Open question: should we go in the rendered order or the layer order? Sometimes labels and 3d buildings are rendered out of order.

## Dispatched once per source feature or once per rendered feature?

If a feature has both a rendered fill and a rendered symbol, and you click on a point covered by both, should two events be dispatched? I think so, because events are added to rendered features (layer ids can be part of the selector).

## Stopping propagation
Should it be possible to prevent other listeners from being called from within an event listener? On the web you can do this with `e.stopPropagation()`.

I think there are definitely use cases where you would want to do this. Like hovering on just the topmost feature:

```
map.on('mouseenter', 'layerid', filter, function(e) {
    // set 'hover' state of feature to true
    ...
    // prevent later mouseenter events from being called on lower features
    e.stopPropagation();
});
map.on('mouseleave', 'layerid', filter, function(e) {
    // set 'hover' state of feature to false
});
```

This adds complexity but seems like something that needs to be done somehow, and I don't have better ideas.

## Summary
So I'm thinking: per-feature event listeners that get individually dispatched, once per rendered feature, in a top-down order with some way of stopping event propagation.

@mollymerp @asheemmamoowala @anandthakker @jfirebaugh @lucaswoj

**Comment by stevage** [comment]
>If a feature has both a rendered fill and a rendered symbol, and you click on a point covered by both, should two events be dispatched?

My instinctive reaction here is no. If I'm attaching a per-feature handler (the primary use case for which would be "tell me every time the mouse moves over a feature"), I can't think of any reason I'd want to get separate events for the two symbols of the same feature. In some cases it would make this handler hard to use.

(Obviously just speaking for the kinds of maps I make.)

>Should it be possible to prevent other listeners from being called from within an event listener?

I'm not sure. As soon as things get that complicated, I'd probably revert to what I do now which is along these lines:

1. Attach a generic click/mousemove handler
2. Call queryRenderedFeatures() on my most important layer.
3. If it, handle that and stop.
4. Call queryRenderedFeatures() on my most second-most important layer.
5. Etc...

In other words: the workaround without such a feature is just fine for me, I think.

**Comment by everhardt** [comment]
@stevage: How would you suggest to do "what you are doing now" when I want to catch `mouseenter` events on only the top most layer? A generic `mouseenter`-handler won't be called when moving around the map.
